presto-spark-base/src/main/java/com/facebook/presto/spark/execution/AbstractPrestoSparkQueryExecution.java

```diff
@@ -708,6 +717,144 @@ else if (writerTarget instanceof ExecutionWriterTarget.RefreshMaterializedViewHa
         }
     }
 
+    // Returns RDD for specified fragmented subplan
+    // This method ensures that RDD is created only once for a sub-plan, where identity is determined by fragment id
+    protected <T extends PrestoSparkTaskOutput> RddAndMore<T> createRddFragmented(SubPlan subPlan, TableWriteInfo tableWriteInfo)
+            throws SparkException, TimeoutException
+    {
+        synchronized (fragmentIdToRdd) {
+            if (fragmentIdToRdd.containsKey(subPlan.getFragment().getId())) {
+                return fragmentIdToRdd.get(subPlan.getFragment().getId());
+            }
+
+            ImmutableMap.Builder<PlanFragmentId, JavaPairRDD<MutablePartitionId, PrestoSparkMutableRow>> rddInputs = ImmutableMap.builder();
+            ImmutableMap.Builder<PlanFragmentId, Broadcast<?>> broadcastInputs = ImmutableMap.builder();
+            ImmutableList.Builder<PrestoSparkBroadcastDependency<?>> broadcastDependencies = ImmutableList.builder();
+
+            for (SubPlan child : subPlan.getChildren()) {
+                RddAndMore<?> childRdd = createRddFragmented(child, tableWriteInfo);
+                if (childRdd.getPartitioningHandle().isPresent() && childRdd.getPartitioningHandle().get().equals(FIXED_BROADCAST_DISTRIBUTION)) {
+                    PrestoSparkBroadcastDependency<?> broadcastDependency = createBroadcastDependency(childRdd);
+                    broadcastInputs.put(child.getFragment().getId(), broadcastDependency.executeBroadcast(sparkContext));
+                    broadcastDependencies.add(broadcastDependency);
+                }
+                else {
+                    rddInputs.put(child.getFragment().getId(), (JavaPairRDD<MutablePartitionId, PrestoSparkMutableRow>) childRdd.getRdd());
+                    broadcastDependencies.addAll(childRdd.getBroadcastDependencies());
+                }
+            }
+
+            Class outputType = PrestoSparkSerializedPage.class;
+            if (!isRootSubPlan(subPlan)) {
```

Comment:
What happens if it is a root subplan?

Reply:
Added explicit if condition. It defaults to PrestoSparkSerializedPage